File: backend/app/utils/job_status.py
"""
Job Status Normalization Utilities

Provides a single source of truth for job status normalization across
the entire application. All endpoints and services should use these
functions to ensure consistent status values.

Canonical status values:
- queued: Job is waiting to be processed
- running: Job is currently being executed
- succeeded: Job completed successfully
- failed: Job encountered an error
- cancelled: Job was cancelled by user
- cancelling: Job cancellation is in progress
"""
from __future__ import annotations

from typing import Any, Dict, Optional


# Canonical status values - use these constants for comparisons
STATUS_QUEUED = "queued"
STATUS_RUNNING = "running"
STATUS_SUCCEEDED = "succeeded"
STATUS_FAILED = "failed"
STATUS_CANCELLED = "cancelled"
STATUS_CANCELLING = "cancelling"
STATUS_PENDING_RETRY = "pending_retry"  # Job failed but will be retried

# All valid terminal statuses (job is done, no more updates expected)
TERMINAL_STATUSES = frozenset({STATUS_SUCCEEDED, STATUS_FAILED, STATUS_CANCELLED})

# All valid active statuses (job may still update)
ACTIVE_STATUSES = frozenset({STATUS_QUEUED, STATUS_RUNNING, STATUS_CANCELLING, STATUS_PENDING_RETRY})

# Statuses that indicate the job will be retried
RETRY_STATUSES = frozenset({STATUS_PENDING_RETRY})


def normalize_job_status(status: Optional[str]) -> str:
    """Normalize job status to consistent canonical values.

    This is the single source of truth for status normalization.
    All endpoints and services should use this function.

    Args:
        status: Raw status string from any source

    Returns:
        Canonical status string (queued, running, succeeded, failed, cancelled, cancelling)

    Examples:
        >>> normalize_job_status("succeeded")
        'succeeded'
        >>> normalize_job_status("completed")
        'succeeded'
        >>> normalize_job_status("in_progress")
        'running'
        >>> normalize_job_status(None)
        'queued'
    """
    value = (status or "").strip().lower()

    # Map to canonical 'succeeded'
    if value in {"succeeded", "success", "done", "completed"}:
        return STATUS_SUCCEEDED

    # Map to canonical 'queued'
    if value in {"queued", "pending", "waiting"}:
        return STATUS_QUEUED

    # Map to canonical 'running'
    if value in {"running", "in_progress", "started", "processing"}:
        return STATUS_RUNNING

    # Map to canonical 'failed'
    if value in {"failed", "error", "errored"}:
        return STATUS_FAILED

    # Map to canonical 'cancelled'
    if value in {"cancelled", "canceled"}:
        return STATUS_CANCELLED

    # Preserve 'cancelling' as-is
    if value == "cancelling":
        return STATUS_CANCELLING

    # Map to canonical 'pending_retry'
    if value in {"pending_retry", "retry_pending", "retry_scheduled", "awaiting_retry"}:
        return STATUS_PENDING_RETRY

    # Default to queued for unknown/empty statuses
    return STATUS_QUEUED


def normalize_job(job: Optional[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    """Normalize a job record for consistent API responses.

    Args:
        job: Raw job dictionary from storage

    Returns:
        Job dictionary with normalized status field
    """
    if not job:
        return job

    normalized = dict(job)

    # Normalize status field
    if "status" in normalized:
        normalized["status"] = normalize_job_status(normalized["status"])
    elif "state" in normalized:
        # Some older records use 'state' instead of 'status'
        normalized["status"] = normalize_job_status(normalized["state"])

    return normalized


def is_terminal_status(status: Optional[str]) -> bool:
    """Check if a status indicates the job is complete (no more updates).

    Args:
        status: Status string (will be normalized)

    Returns:
        True if the job is in a terminal state
    """
    return normalize_job_status(status) in TERMINAL_STATUSES


def is_active_status(status: Optional[str]) -> bool:
    """Check if a status indicates the job may still be updated.

    Args:
        status: Status string (will be normalized)

    Returns:
        True if the job is still active
    """
    return normalize_job_status(status) in ACTIVE_STATUSES


def is_pending_retry(status: Optional[str]) -> bool:
    """Check if a status indicates the job is waiting for retry.

    Args:
        status: Status string (will be normalized)

    Returns:
        True if the job is pending retry
    """
    return normalize_job_status(status) in RETRY_STATUSES


def can_retry(job: Optional[Dict[str, Any]]) -> bool:
    """Check if a job can be retried.

    A job can be retried if:
    - It exists
    - Its status is 'failed' (not pending_retry, which auto-retries)
    - It has not exceeded max retries

    Args:
        job: Job dictionary from storage

    Returns:
        True if the job can be manually retried
    """
    if not job:
        return False

    status = normalize_job_status(job.get("status"))
    if status != STATUS_FAILED:
        return False

    # Check retry count vs max retries
    retry_count = job.get("retryCount") or job.get("retry_count") or 0
    max_retries = job.get("maxRetries") or job.get("max_retries") or 3

    return retry_count < max_retries

File: frontend/src/utils/jobStatus.js
/**
 * Centralized Job Status Utilities
 *
 * This module provides consistent job status normalization and display
 * configuration across all components.
 */

/**
 * Canonical job status values used throughout the application.
 * Backend may return different values (succeeded, queued) which get normalized.
 */
export const JobStatus = {
  PENDING: 'pending',
  RUNNING: 'running',
  COMPLETED: 'completed',
  FAILED: 'failed',
  CANCELLED: 'cancelled',
  CANCELLING: 'cancelling',
}

/**
 * Status values considered "active" (job is still processing)
 */
export const ACTIVE_STATUSES = new Set([JobStatus.PENDING, JobStatus.RUNNING, JobStatus.CANCELLING])

/**
 * Status values considered "terminal" (job has finished)
 */
export const TERMINAL_STATUSES = new Set([
  JobStatus.COMPLETED,
  JobStatus.FAILED,
  JobStatus.CANCELLED,
])

/**
 * Status values considered "failure" states
 */
export const FAILURE_STATUSES = new Set([JobStatus.FAILED, JobStatus.CANCELLED])

/**
 * Normalize a job status string to a canonical value.
 * Handles various backend representations (succeeded, queued, etc.)
 *
 * @param {string|undefined|null} status - Raw status value
 * @returns {string} Normalized status
 */
export function normalizeJobStatus(status) {
  const value = (status || '').toString().toLowerCase().trim()

  // Map backend "succeeded" to frontend "completed"
  if (value === 'succeeded' || value === 'success' || value === 'done') {
    return JobStatus.COMPLETED
  }

  // Map "queued" to "pending" for consistent display
  if (value === 'queued') {
    return JobStatus.PENDING
  }

  // Handle in_progress variant
  if (value === 'in_progress' || value === 'started') {
    return JobStatus.RUNNING
  }

  // Handle error variant
  if (value === 'error') {
    return JobStatus.FAILED
  }

  // Handle canceled spelling variant
  if (value === 'canceled') {
    return JobStatus.CANCELLED
  }

  // Handle cancelling state
  if (value === 'canceling') {
    return JobStatus.CANCELLING
  }

  // Return as-is if it's a known canonical status
  if (Object.values(JobStatus).includes(value)) {
    return value
  }

  // Default to pending for unknown statuses
  return JobStatus.PENDING
}

/**
 * Normalize step status (may have different values than job status)
 *
 * @param {string|undefined|null} status - Raw step status
 * @returns {string} Normalized status
 */
export function normalizeStepStatus(status) {
  const value = (status || '').toString().toLowerCase().trim()

  // Steps may report "complete" instead of "completed"
  if (value === 'complete' || value === 'done' || value === 'success') {
    return JobStatus.COMPLETED
  }

  // Steps report "skipped" which we treat as completed
  if (value === 'skipped') {
    return JobStatus.COMPLETED
  }

  return normalizeJobStatus(value)
}

/**
 * Check if a job status indicates the job is still active/processing
 *
 * @param {string} status - Job status (can be raw or normalized)
 * @returns {boolean}
 */
export function isActiveStatus(status) {
  return ACTIVE_STATUSES.has(normalizeJobStatus(status))
}

/**
 * Check if a job status indicates the job has completed (successfully or not)
 *
 * @param {string} status - Job status (can be raw or normalized)
 * @returns {boolean}
 */
export function isTerminalStatus(status) {
  return TERMINAL_STATUSES.has(normalizeJobStatus(status))
}

/**
 * Check if a job status indicates a failure
 *
 * @param {string} status - Job status (can be raw or normalized)
 * @returns {boolean}
 */
export function isFailureStatus(status) {
  return FAILURE_STATUSES.has(normalizeJobStatus(status))
}

/**
 * Check if a job can be retried (must be in a failure state)
 *
 * @param {string} status - Job status
 * @returns {boolean}
 */
export function canRetryJob(status) {
  return normalizeJobStatus(status) === JobStatus.FAILED
}

/**
 * Check if a job can be cancelled (must be active)
 *
 * @param {string} status - Job status
 * @returns {boolean}
 */
export function canCancelJob(status) {
  return isActiveStatus(status)
}

/**
 * Get human-readable label for a status
 *
 * @param {string} status - Job status (can be raw or normalized)
 * @returns {string}
 */
export function getStatusLabel(status) {
  const normalized = normalizeJobStatus(status)
  const labels = {
    [JobStatus.PENDING]: 'Pending',
    [JobStatus.RUNNING]: 'Running',
    [JobStatus.COMPLETED]: 'Completed',
    [JobStatus.FAILED]: 'Failed',
    [JobStatus.CANCELLED]: 'Cancelled',
    [JobStatus.CANCELLING]: 'Cancelling',
  }
  return labels[normalized] || 'Unknown'
}

/**
 * Normalize a job object, converting all status-related fields
 *
 * @param {Object} job - Raw job object from API
 * @returns {Object} Normalized job object
 */
export function normalizeJob(job = {}) {
  const status = normalizeJobStatus(job.status || job.state)
  const result = job.result || {}
  const meta = job.meta || job.metadata || {}

  const artifacts = job.artifacts || result.artifacts || {
    html_url: result.html_url,
    pdf_url: result.pdf_url,
    docx_url: result.docx_url,
    xlsx_url: result.xlsx_url,
  }

  // Normalize step statuses if present
  const steps = Array.isArray(job.steps)
    ? job.steps.map(step => ({
        ...step,
        status: normalizeStepStatus(step.status),
      }))
    : job.steps

  return {
    ...job,
    status,
    steps,
    templateName: job.templateName || job.template_name || job.template || job.templateTitle,
    templateId: job.templateId || job.template_id,
    templateKind: job.templateKind || job.template_kind || job.kind,
    connectionId: job.connectionId || job.connection_id,
    createdAt: job.createdAt || job.created_at || job.startedAt || job.started_at || job.queuedAt || job.queued_at,
    startedAt: job.startedAt || job.started_at || job.created_at || job.queuedAt || job.queued_at,
    finishedAt: job.finishedAt || job.finished_at || job.completed_at || job.completedAt,
    startDate: meta.start_date || meta.startDate || job.start_date || job.startDate,
    endDate: meta.end_date || meta.endDate || job.end_date || job.endDate,
    keyValues: meta.key_values || meta.keyValues || job.key_values || job.keyValues,
    artifacts,
    meta,
  }
}
